file_sync_service.c

```diff
@@ -534,45 +534,37 @@ static int state_process_send_header(afs_service_t *svc, apacket *p)
 static int state_process_send_file(afs_service_t *svc, apacket *p)
 {
     int ret;
-
-    while (svc->namelen > 0) {
-        int block_size = min(SYNC_TEMP_BUFF_SIZE, svc->namelen);
-
-        ret = read_from_packet(svc, p, block_size);
-        if (ret != 0) {
-            if (ret == -EAGAIN) {
-                return 1;
+    int block_size = min(p->msg.data_length, svc->namelen);
+    uint8_t *write_ptr = svc->packet_ptr;
+
+    svc->namelen -= block_size;
+    p->msg.data_length -= block_size;
+    svc->packet_ptr += block_size;
```

Comment:
IIUC I see two changes here:
- `while` loop is removed to rely on the caller's one
- `read_from_packet()` is removed to optimize the `write()` without going through the `memcpy()`. For this change I would add another function or macro to handle in the same place the `packet_ptr` and `p->msgdata_length` logic. I'll merge this PR and think about this.

Reply:
Sure.